### PR TITLE
[PASS] add a pass for the specific hardware accelarator when it is not binded

### PR DESCRIPTION
**Pull request by kun-zh**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

For accelerators, host device splitting is different from gpu, as discussed here, 
https://discuss.tvm.ai/t/split-we-want-to-split-the-host-and-device-code-flexibly/932

this is a solution we are using for DaVinci core, and can use for other accelerators.

@tqchen @yzhliu @tmoreau89 @ZihengJiang please review, thanks!

### CODE REVIEW

**Comment by yzhliu** [comment]
Thanks for contribution. Could you add unittest for this?

File: tests/python/unittest/test_pass_detect_device.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    s = tvm.create_schedule(A2.op)
+    xo, xi = s[A2].split(A2.op.axis[0], factor=8)
+    s[A1].compute_at(s[A2], xo)
+    s[A1].set_scope("shared")
+
+    bounds = tvm.schedule.InferBound(s)
+    stmt = tvm.schedule.ScheduleOps(s, bounds)
+    stmt = tvm.ir_pass.DeviceMark(stmt)
+    print stmt
```
Comment (ZihengJiang) [inline tests/python/unittest/test_pass_detect_device.py]: try to use `assert` instead of `print`, you can find some examples in other tests

**Comment by kun-zh** [comment]
add the unit case @yzhliu

File: include/tvm/ir_pass.h [hunk 2]
```diff
@@ -326,6 +326,15 @@ Stmt RewriteUnsafeSelect(Stmt stmt);
...
 Stmt LowerStorageAccessInfo(Stmt stmt);
 
+/*!
+ * \brief insert the mark of device for the hardware accelarator when 
+ * it is not binded with thread or block.
+ *
+ * \param stmt The stmt to be trasnformed
+ * \return Transformed stmt.
+ */
+Stmt DeviceMark(Stmt stmt);
```
Comment (tqchen) [inline include/tvm/ir_pass.h]: as per our current convention, this should be Mark Device

File: src/pass/split_host_device.cc [hunk 3]
```diff
@@ -153,7 +153,9 @@ class HostDeviceSplitter : public IRMutator {
 
   Stmt Mutate_(const AttrStmt *op, const Stmt& s) final {
     if (op->attr_key == attr::thread_extent ||
-        op->attr_key == attr::pipeline_exec_scope) {
+        op->attr_key == attr::pipeline_exec_scope ||
+        (op->attr_key == attr::pragma_scope_prefix &&
+        op->value.as<StringImm>()->value == "device")) {
```
Comment (tqchen) [inline src/pass/split_host_device.cc]: in our current convention, we use attr=pragma_xyz to indicate a special pragma, so if we want to do this, we should use pragma_device in the key and ignore the value(can be just constant 0)

File: src/pass/detect_device.cc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#include "../pass/ir_util.h"
+
+namespace tvm {
+namespace ir {
+
+class DetectDevice : public IRMutator {
+ public:
+  DetectDevice() {}
+  Stmt Detect(Stmt stmt) {
+    Stmt body = AttrStmt::make(make_zero(Int(32)),
```
Comment (tqchen) [inline src/pass/detect_device.cc]: Seems that this does not do anything except for the wrapping, is it intentional? The simple wrapping can be constructed by directly use AttrStmt constructor and does not seems to need a pass

**Comment by kun-zh** [comment]
Actually, it is a feature for some hardware accelerator which can't be binded. The pass is not essential for GPU or CPU. Using a pass will be flexible, user can add it in their own custom pass list, just like coproc_sync. And there will be no side-effect for current solution. @tqchen

File: src/pass/detect_device.cc [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <tvm/ir_pass.h>
+#include <tvm/ir_mutator.h>
+#include "../pass/ir_util.h"
+
+namespace tvm {
+namespace ir {
+
+class DetectDevice : public IRMutator {
+ public:
```
Comment (tqchen) [inline src/pass/detect_device.cc]: IRMutator is not necessary, if we want to support this one, we can just inline the logic to MarkDevice

File: src/pass/detect_device.cc [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  Stmt Detect(Stmt stmt) {
+    Stmt body = AttrStmt::make(make_zero(Int(32)),
+                               ir::attr::device_scope,
+                               0,
+                               stmt);
+    return body;
+  }
+};
+
+Stmt MarkDevice(Stmt stmt) {
```
Comment (tqchen) [inline src/pass/detect_device.cc]: MarkDevice->DecorateDeviceScope, which reflects what we really want

File: include/tvm/ir_pass.h [hunk 7]
```diff
@@ -326,6 +326,15 @@ Stmt RewriteUnsafeSelect(Stmt stmt);
  */
 Stmt LowerStorageAccessInfo(Stmt stmt);
 
+/*!
+ * \brief insert the mark of device for the hardware accelarator when
```
Comment (tqchen) [inline include/tvm/ir_pass.h]: decorate the stmt with a device scope, this is helpful for hardware accelerator without thread blocks

**Comment by kun-zh** [comment]
@tqchen Changed per the request

**Comment by tqchen** [comment]
Thanks @kun-zh @ZihengJiang , this is merged

**Pull request merged**